File: packages/core/src/social/log.ts
/**
 * Log operations for GitSocial - Direct git access for complete audit trail
 *
 * Uses direct git operations to retrieve ALL git history including:
 * - Regular commits (implicit posts)
 * - GitMsg posts, comments, reposts, quotes
 * - GitSocial metadata operations (lists, config)
 * - refs/gitmsg/* operations
 */

import type { LogEntry, Result } from './types';
import type { Commit } from '../git/types';
import type { GitMsgMessage } from '../gitmsg/types';
import { git } from '../git';
import { gitMsg } from '../gitmsg';
import { gitMsgRef, gitMsgUrl } from '../gitmsg/protocol';
import { storage } from '../storage';
import { existsSync } from 'fs';
import { log as logger } from '../logger';

/**
 * Log namespace - Log management operations
 */
export const log = {
  getLogs,
  initialize
};

/**
 * Initialize the logs system with storage configuration
 * @deprecated Storage base should be passed via getLogs options
 */
function initialize(_config: { storageBase: string }): void {
  logger('debug', '[social.logs] Initialize called (deprecated - pass storageBase via getLogs options)');
}

async function getLogs(
  workdir: string,
  scope: string = 'repository:my',
  filter?: {
    since?: Date;
    until?: Date;
    limit?: number;
    types?: Array<LogEntry['type']>;
    storageBase?: string;
  }
): Promise<Result<LogEntry[]>> {
  try {
    logger('debug', '[social.log] getLogs called with:', {
      scope,
      since: filter?.since,
      until: filter?.until,
      limit: filter?.limit
    });

    // Parse scope to determine data source
    if (scope.startsWith('repository:') && scope !== 'repository:my') {
      // External repository - use storage layer
      logger('debug', '[social.log] Using external repository logs for scope:', scope);
      return await getExternalRepositoryLogs(scope, filter);
    } else {
      // Workspace repository - use local git operations
      logger('debug', '[social.log] Using workspace repository logs for scope:', scope);
      return await getWorkspaceRepositoryLogs(workdir, scope, filter);
    }
  } catch (error) {
    return {
      success: false,
      error: {
        code: 'GET_LOGS_ERROR',
        message: error instanceof Error ? error.message : 'Failed to get logs'
      }
    };
  }
}

async function getExternalRepositoryLogs(
  scope: string,
  filter?: {
    since?: Date;
    until?: Date;
    limit?: number;
    types?: Array<LogEntry['type']>;
    storageBase?: string;
  }
): Promise<Result<LogEntry[]>> {
  // Extract repository URL from scope (e.g., 'repository:https://github.com/user/repo#branch:main')
  const repositoryUrl = scope.replace('repository:', '');

  // Use storage layer to get commits from external repository

  // Parse repository URL and branch
  const parsed = gitMsgRef.parseRepositoryId(repositoryUrl);
  if (!parsed) {
    throw new Error(`Invalid repository URL in scope: ${repositoryUrl}`);
  }

  const effectiveStorageBase = filter?.storageBase;
  if (!effectiveStorageBase) {
    logger('error', '[social.log] No storage base provided for external repository logs');
    return {
      success: false,
      error: {
        code: 'NO_STORAGE_BASE',
        message: 'Storage base not provided for external repository logs'
      }
    };
  }
  const storageDir = storage.path.getDirectory(effectiveStorageBase, parsed.repository);
  logger('debug', '[social.log] External repository storage dir:', storageDir);

  if (!existsSync(storageDir)) {
    return {
      success: false,
      error: {
        code: 'REPOSITORY_NOT_CLONED',
        message: `Repository not cloned: ${repositoryUrl}`
      }
    };
  }

  // Get commits using existing storage infrastructure
  const result = await storage.repository.getCommits(effectiveStorageBase, parsed.repository, {
    branch: parsed.branch,
    limit: filter?.limit,
    since: filter?.since,
    until: filter?.until
  });

  if (!result.success || !result.data) {
    return {
      success: false,
      error: result.error || {
        code: 'GET_COMMITS_FAILED',
        message: 'Failed to get commits from external repository'
      }
    };
  }

  const logEntries = await transformCommitsToLogEntries(result.data, repositoryUrl);

  logger('debug', '[social.log] Transformed commits to log entries:', {
    commitCount: result.data.length,
    logEntryCount: logEntries.length,
    dateRange: {
      since: filter?.since,
      until: filter?.until,
      firstEntry: logEntries[0]?.timestamp,
      lastEntry: logEntries[logEntries.length - 1]?.timestamp
    }
  });

  // Apply type filter if specified
  const filtered = filter?.types
    ? logEntries.filter(entry => filter.types!.includes(entry.type))
    : logEntries;

  return { success: true, data: filtered };
}

async function getWorkspaceRepositoryLogs(
  workdir: string,
  scope: string,
  filter?: {
    since?: Date;
    until?: Date;
    limit?: number;
    types?: Array<LogEntry['type']>;
    storageBase?: string;
  }
): Promise<Result<LogEntry[]>> {
  const gitSocialBranch = await git.getConfiguredBranch(workdir);

  // Get all refs/gitmsg/* refs for metadata operations
  const refs = await git.listRefs(workdir, 'social');
  const refPaths = refs.map((ref: string) => `refs/gitmsg/${ref}`);

  let commits: Commit[] = [];

  if (scope === 'repository:my') {
    // Repository scope: local branches + origin + GitSocial refs
    commits = await git.getCommits(workdir, {
      all: false,
      branch: gitSocialBranch,
      includeRefs: refPaths,
      since: filter?.since,
      until: filter?.until,
      limit: filter?.limit
    });
  } else if (scope === 'timeline') {
    // Timeline scope: all branches including remotes + GitSocial refs
    commits = await git.getCommits(workdir, {
      all: true,
      since: filter?.since,
      until: filter?.until,
      limit: filter?.limit
    });
  }

  // Transform commits to log entries
  const logEntries = await transformCommitsToLogEntries(commits, workdir);

  // Apply type filter if specified
  const filtered = filter?.types
    ? logEntries.filter(entry => filter.types!.includes(entry.type))
    : logEntries;

  return { success: true, data: filtered };
}

async function transformCommitsToLogEntries(commits: Commit[], repository: string): Promise<LogEntry[]> {
  const logEntries: LogEntry[] = [];

  // Build ref map for metadata operations
  const refMap = new Map<string, string>();
  for (const commit of commits) {
    if (commit.refname && commit.refname.startsWith('refs/gitmsg/')) {
      const ref = commit.refname.replace(/^refs\/gitmsg\//, '');
      refMap.set(commit.hash, ref);
    }
  }

  // Transform each commit
  for (const commit of commits) {
    try {
      const actionType = getActionType(commit, refMap);
      const gitMsgParsed = gitMsg.parseMessage(commit.message);

      // Handle list state changes specially - generate virtual follow/unfollow entries
      if (actionType === 'list-state-change') {
        const ref = refMap.get(commit.hash);
        if (ref && ref.startsWith('social/lists/')) {
          const listName = ref.replace('social/lists/', '');
          const virtualEntries = await generateVirtualRepositoryEntries(commit, ref, listName, repository);
          logEntries.push(...virtualEntries);
        }
      } else {
        // Regular log entry
        // Parse repository to get base URL without branch
        const parsedRepo = gitMsgRef.parseRepositoryId(repository);
        const baseRepository = parsedRepo?.repository || repository;

        const logEntry: LogEntry = {
          hash: commit.hash.substring(0, 12),
          timestamp: commit.timestamp,
          author: {
            name: commit.author,
            email: commit.email
          },
          type: actionType as LogEntry['type'],
          details: formatActionDetails(actionType, gitMsgParsed, commit),
          repository,
          postId: ['post', 'comment', 'repost', 'quote'].includes(actionType)
            ? gitMsgRef.create('commit', commit.hash, gitMsgUrl.validate(baseRepository) ? baseRepository : undefined)
            : undefined,
          raw: { commit, gitMsg: gitMsgParsed || undefined }
        };

        logEntries.push(logEntry);
      }
    } catch (error) {
      logger('warn', '[social.getLogs] Error processing commit for logs:', error);
      // Skip invalid commits
    }
  }

  // Sort by timestamp (newest first)
  logEntries.sort((a, b) => b.timestamp.getTime() - a.timestamp.getTime());

  return logEntries;
}

function getActionType(commit: Commit, refMap: Map<string, string>): string {
  // Check if this commit is pointed to by any refs/gitmsg/* refs
  const ref = refMap.get(commit.hash);

  if (ref) {
    // This commit is a special GitSocial metadata operation
    if (ref.startsWith('social/lists/')) {
      // List operation - state-based storage, will generate virtual follow/unfollow events
      return 'list-state-change';
    } else if (ref === 'social/config') {
      return 'config';
    } else {
      return 'metadata';
    }
  }

  // Check GitMsg headers for interaction types (comment, repost, quote)
  const gitMsgParsed = gitMsg.parseMessage(commit.message);
  if (gitMsgParsed && gitMsgParsed.header.ext === 'social') {
    const socialType = gitMsgParsed.header.fields['type'];
    if (socialType && ['comment', 'repost', 'quote'].includes(socialType)) {
      return socialType;
    }
  }

  // Default to 'post' for commits without GitMsg headers (implicit posts)
  return 'post';
}

function formatActionDetails(type: string, gitMsg: GitMsgMessage | null, commit: Commit): string {
  // Use commit message as primary content source
  const primaryContent = commit.message?.split('\n')[0] || 'No message';

  switch (type) {
  case 'list-create': {
    // For list operations, the commit message contains JSON data directly
    try {
      const listData = JSON.parse(commit.message) as { name?: string; id?: string };
      return `Created list "${listData.name || listData.id || 'unknown'}"`;
    } catch {
      return 'Created list "unknown"';
    }
  }
  case 'list-delete':
    return 'Deleted list';
  case 'repository-follow': {
    const repo = gitMsg?.header?.fields['social:repository'] || 'unknown';
    const listName = gitMsg?.header?.fields['social:list'] || 'unknown';
    return `Added ${repo} to list "${listName}"`;
  }
  case 'repository-unfollow': {
    const repo = gitMsg?.header?.fields['social:repository'] || 'unknown';
    const listName = gitMsg?.header?.fields['social:list'] || 'unknown';
    return `Removed ${repo} from list "${listName}"`;
  }
  case 'config':
    return gitMsg?.content || primaryContent;
  case 'metadata':
    return gitMsg?.content || primaryContent;
  case 'post':
    // For posts, prefer GitMsg content if available, otherwise use commit message
    return gitMsg?.content?.split('\n')[0] || primaryContent;
  case 'comment': {
    const original = gitMsg?.references?.find((ref) => ref.fields['social:ref-type'] === 'original');
    const originalContent = original?.metadata?.split('\n')?.find((line: string) => line.startsWith('>'))?.substring(1)?.trim();
    return `Re: ${originalContent || 'post'}`;
  }
  case 'repost': {
    const original = gitMsg?.references?.find((ref) => ref.fields['social:ref-type'] === 'original');
    const originalContent = original?.metadata?.split('\n')?.find((line: string) => line.startsWith('>'))?.substring(1)?.trim();
    return `Repost: ${originalContent || 'post'}`;
  }
  case 'quote': {
    const original = gitMsg?.references?.find((ref) => ref.fields['social:ref-type'] === 'original');
    const originalContent = original?.metadata?.split('\n')?.find((line: string) => line.startsWith('>'))?.substring(1)?.trim();
    return `Quote: ${originalContent || 'post'}`;
  }
  default:
    return primaryContent;
  }
}

async function getPreviousListCommit(currentCommit: Commit, ref: string, workdir: string): Promise<Commit | null> {
  try {
    // Get the previous commit for this specific ref
    const refPath = `refs/gitmsg/${ref}`;
    const commits = await git.getCommits(workdir, {
      branch: refPath,
      limit: 2
    });

    // Find the commit before the current one
    return commits.find(c => c.hash !== currentCommit.hash) || null;
  } catch (error) {
    logger('debug', '[getPreviousListCommit] Failed to get previous commit:', error);
    return null;
  }
}

function parseListState(commit: Commit): { name?: string; repositories: string[] } {
  try {
    const listData = JSON.parse(commit.message) as { name?: string; repositories?: string[] };
    return {
      name: listData.name,
      repositories: listData.repositories || []
    };
  } catch (error) {
    logger('debug', '[parseListState] Failed to parse list state:', error);
    return { repositories: [] };
  }
}

function calculateRepositoryDiff(previous: string[], current: string[]): {
  added: string[];
  removed: string[];
} {
  const previousSet = new Set(previous);
  const currentSet = new Set(current);

  const added = current.filter(repo => !previousSet.has(repo));
  const removed = previous.filter(repo => !currentSet.has(repo));

  return { added, removed };
}

async function generateVirtualRepositoryEntries(
  commit: Commit,
  ref: string,
  listName: string,
  repository: string
): Promise<LogEntry[]> {
  const entries: LogEntry[] = [];

  try {
    // Parse current list state
    const currentState = parseListState(commit);

    // Get previous commit and parse its state
    const previousCommit = await getPreviousListCommit(commit, ref, repository);
    const previousState = previousCommit ? parseListState(previousCommit) : { repositories: [] };

    // Calculate what changed
    const { added, removed } = calculateRepositoryDiff(previousState.repositories, currentState.repositories);

    // Generate virtual entries for added repositories (follow events)
    for (const repoUrl of added) {
      entries.push({
        hash: commit.hash.substring(0, 12),
        timestamp: commit.timestamp,
        author: {
          name: commit.author,
          email: commit.email
        },
        type: 'repository-follow',
        details: `Added ${repoUrl} to list "${listName}"`,
        repository,
        raw: { commit, gitMsg: undefined }
      });
    }

    // Generate virtual entries for removed repositories (unfollow events)
    for (const repoUrl of removed) {
      entries.push({
        hash: commit.hash.substring(0, 12),
        timestamp: commit.timestamp,
        author: {
          name: commit.author,
          email: commit.email
        },
        type: 'repository-unfollow',
        details: `Removed ${repoUrl} from list "${listName}"`,
        repository,
        raw: { commit, gitMsg: undefined }
      });
    }

    // If no repositories changed, show it as a list update
    if (added.length === 0 && removed.length === 0) {
      entries.push({
        hash: commit.hash.substring(0, 12),
        timestamp: commit.timestamp,
        author: {
          name: commit.author,
          email: commit.email
        },
        type: 'list-create',
        details: `Updated list "${listName}"`,
        repository,
        raw: { commit, gitMsg: undefined }
      });
    }

  } catch (error) {
    logger('warn', '[generateVirtualRepositoryEntries] Error generating virtual entries:', error);
    // Fallback to basic list entry
    entries.push({
      hash: commit.hash.substring(0, 12),
      timestamp: commit.timestamp,
      author: {
        name: commit.author,
        email: commit.email
      },
      type: 'list-create',
      details: `Updated list "${listName}"`,
      repository,
      raw: { commit, gitMsg: undefined }
    });
  }

  return entries;
}

File: library/extensions/social/log.go
// log.go - Activity log retrieval and formatting
package social

import (
	"strings"
	"time"

	"github.com/gitsocial-org/gitsocial/library/core/git"
	"github.com/gitsocial-org/gitsocial/library/core/gitmsg"
	"github.com/gitsocial-org/gitsocial/library/core/log"
	"github.com/gitsocial-org/gitsocial/library/core/protocol"
	"github.com/gitsocial-org/gitsocial/library/core/text"
)

type GetLogsOptions struct {
	Limit  int
	Types  []LogEntryType
	After  *time.Time
	Before *time.Time
	Author string
}

// GetLogs retrieves activity log entries with optional filtering.
func GetLogs(workdir, scope string, opts *GetLogsOptions) Result[[]LogEntry] {
	if opts == nil {
		opts = &GetLogsOptions{Limit: 20}
	}
	if opts.Limit == 0 {
		opts.Limit = 20
	}

	branch := gitmsg.GetExtBranch(workdir, "social")

	gitOpts := &git.GetCommitsOptions{
		Branch: branch,
		Limit:  opts.Limit * 2,
	}

	switch {
	case scope == "" || scope == "repository:my":
		// default: current workspace
	case scope == "timeline":
		gitOpts.All = true
	case strings.HasPrefix(scope, "list:"):
		return failure[[]LogEntry]("INVALID_SCOPE", "list scope is not supported for logs: use gitsocial search")
	case strings.HasPrefix(scope, "repository:"):
		return failure[[]LogEntry]("INVALID_SCOPE", "external repository scope is not supported for logs: use gitsocial search")
	default:
		return failure[[]LogEntry]("INVALID_SCOPE", "unknown scope: "+scope)
	}

	commits, err := git.GetCommits(workdir, gitOpts)
	if err != nil {
		return failureWithDetails[[]LogEntry]("GIT_ERROR", "read commits", err)
	}

	refs, err := git.ListRefs(workdir, "social/")
	if err != nil {
		log.Debug("list social refs failed", "error", err)
	}
	refMap := make(map[string]string)
	for _, ref := range refs {
		hash, err := git.ReadRef(workdir, "refs/gitmsg/"+ref)
		if err != nil {
			log.Debug("read social ref failed", "ref", ref, "error", err)
			continue
		}
		if hash != "" {
			refMap[strings.TrimSpace(hash)] = ref
		}
	}

	var entries []LogEntry
	for _, commit := range commits {
		entry := commitToLogEntry(commit, refMap)
		if !matchesLogFilters(entry, opts) {
			continue
		}
		entries = append(entries, entry)
		if len(entries) >= opts.Limit {
			break
		}
	}

	return success(entries)
}

// matchesLogFilters checks if a log entry matches the filter criteria.
func matchesLogFilters(entry LogEntry, opts *GetLogsOptions) bool {
	if len(opts.Types) > 0 {
		found := false
		for _, t := range opts.Types {
			if t == entry.Type {
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}
	if opts.Author != "" && !strings.Contains(strings.ToLower(entry.Author.Email), strings.ToLower(opts.Author)) {
		return false
	}
	if opts.After != nil && entry.Timestamp.Before(*opts.After) {
		return false
	}
	if opts.Before != nil && entry.Timestamp.After(*opts.Before) {
		return false
	}
	return true
}

// commitToLogEntry converts a git commit to a LogEntry.
func commitToLogEntry(commit git.Commit, refMap map[string]string) LogEntry {
	msg := protocol.ParseMessage(commit.Message)
	entryType := detectLogEntryType(commit, msg, refMap)
	details := formatLogDetails(commit, msg, entryType)

	// An activity entry names who wrote the item: the origin author and time over the importer's.
	var header *protocol.Header
	if msg != nil {
		header = &msg.Header
	}
	authorName, authorEmail := protocol.EffectiveAuthor(header, commit.Author, commit.Email)
	return LogEntry{
		Hash:      commit.Hash,
		Timestamp: protocol.EffectiveTime(header, commit.Timestamp),
		Author: Author{
			Name:  authorName,
			Email: authorEmail,
		},
		Type:       entryType,
		Details:    details,
		Repository: extractRepoFromRefname(commit.Refname),
		PostID:     commit.Hash,
	}
}

// detectLogEntryType determines the type of activity from a commit.
func detectLogEntryType(commit git.Commit, msg *protocol.Message, refMap map[string]string) LogEntryType {
	// The ref shapes are the ones git.ListRefs returns, with refs/gitmsg/ trimmed.
	if ref, ok := refMap[commit.Hash]; ok {
		if strings.HasPrefix(ref, "social/lists/") {
			if strings.Contains(commit.Message, "deleted") || strings.Contains(commit.Message, "remove") {
				return logTypeListDelete
			}
			return logTypeListCreate
		}
		if strings.HasPrefix(ref, "social/config") {
			return logTypeConfig
		}
		return logTypeMetadata
	}

	// An entry names the item by its own header type, not by what it references.
	switch getPostType(msg) {
	case PostTypeComment:
		return logTypeComment
	case PostTypeRepost:
		return logTypeRepost
	case PostTypeQuote:
		return logTypeQuote
	}
	return logTypePost
}

// formatLogDetails creates a summary string for a log entry.
func formatLogDetails(commit git.Commit, _ *protocol.Message, entryType LogEntryType) string {
	content := protocol.ExtractCleanContent(commit.Message)
	content = text.Truncate(content, 77)
	content = strings.ReplaceAll(content, "\n", " ")

	switch entryType {
	case logTypeComment:
		return "Re: " + content
	case logTypeRepost:
		return "Repost: " + content
	case logTypeQuote:
		return "Quote: " + content
	case logTypeListCreate:
		return "Created list"
	case logTypeListDelete:
		return "Deleted list"
	case logTypeConfig:
		return "Updated config"
	case logTypeMetadata:
		return "Metadata update"
	default:
		return content
	}
}

// extractRepoFromRefname extracts the remote name from a git refname.
func extractRepoFromRefname(refname string) string {
	if refname == "" {
		return ""
	}
	if strings.HasPrefix(refname, "refs/remotes/") {
		parts := strings.SplitN(refname, "/", 4)
		if len(parts) >= 3 {
			return parts[2]
		}
	}
	return ""
}
